contracts/PillarStaking.sol

```diff
@@ -0,0 +1,271 @@
+    event MaxStakeAmountUpdated(uint256 newMaxStake);
+
+    /* ========== MODIFIERS ========== */
+
+    modifier whenInitialized() {
+        if (state != StakingState.INITIALIZED) revert OnlyWhenInitialized();
+        _;
+    }
+
+    modifier whenStakeable() {
+        if (state != StakingState.STAKEABLE) revert OnlyWhenStakeable();
+        _;
+    }
+
+    modifier whenStaked() {
+        if (state != StakingState.STAKED) revert OnlyWhenStaked();
+        _;
+    }
+
+    modifier whenReadyForUnstake() {
+        if (state != StakingState.READY_FOR_UNSTAKE)
+            revert OnlyWhenReadyForUnstake();
+        _;
+    }
+
+    modifier whenRewardsAllocated() {
+        if (!rewardsAllocated) revert RewardsNotAllocated();
+        _;
+    }
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        address _stakingToken,
+        address _rewardToken,
+        uint256 _maxTotalStake
+    ) {
+        if (_stakingToken == address(0)) revert InvalidStakingToken();
+        if (_rewardToken == address(0)) revert InvalidRewardsToken();
+        if (_maxTotalStake > 0) maxTotalStake = _maxTotalStake;
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+        setStateInitialized();
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function stake(uint256 _amount) external nonReentrant whenStakeable {
+        if (block.timestamp > stakingPeriod + STAKING_PERIOD)
+            revert("StakingPeriodPassed()");
+        if ((totalStaked + _amount) > maxTotalStake)
+            revert MaximumTotalStakeReached({
+                totalMaxStake: maxTotalStake,
+                currentStakedAmount: totalStaked,
+                remainingStakeableAmount: (maxTotalStake - totalStaked),
+                stakerAmount: _amount
+            });
+        if (_amount < minStake)
+            revert InvalidMinimumStake({minimumStakeAmount: minStake});
+        if (_amount > maxStake)
+            revert InvalidMaximumStake({maximumStakeAmount: maxStake});
+        if (_amount > maxStake - stakeholderData[msg.sender].stakedAmount)
+            revert StakeWouldBeGreaterThanMax();
+        if (stakingToken.balanceOf(msg.sender) < _amount)
+            revert InsufficientBalance();
+        totalStaked = totalStaked + _amount;
+        stakeholderData[msg.sender].stakedAmount =
+            stakeholderData[msg.sender].stakedAmount +
+            _amount;
+        if (!stakeholderData[msg.sender].staked) {
+            stakeholderList.push(msg.sender);
+        }
+        stakingToken.safeTransferFrom(msg.sender, address(this), _amount);
+        emit Staked(msg.sender, _amount);
+    }
+
+    function unstake()
+        external
+        nonReentrant
+        whenReadyForUnstake
+        whenRewardsAllocated
+    {
+        uint256 stakedBalance = stakeholderData[msg.sender].stakedAmount;
+        totalStaked = totalStaked - stakedBalance;
```

Comment:
Please use SafeMath to ensure there are no overflows.

Reply:
SafeMath isnt needed in sol v0.8.0+ its in built

Reply:
Ah, I didn't notice your compiler version.